src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -2,25 +2,27 @@
 
 import com.es.phoneshop.model.product.exception.ProductNotFoundException;
 
-import java.math.BigDecimal;
-import java.util.ArrayList;
-import java.util.Currency;
-import java.util.List;
+import java.util.*;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static final ProductDao instance = new ArrayListProductDao();
+
     private final List<Product> products;
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
     private final Lock readLock = lock.readLock();
     private final Lock writeLock = lock.writeLock();
     private long currentId;
 
+    public static synchronized ProductDao getInstance() {
```

Comment:
Please try to implement this as a singleton.

Reply:
Why it is bad to initialize field ProductDao instance along with it declaration

Reply:
Please check the first answer: https://stackoverflow.com/questions/11165852/java-singleton-and-synchronization